make our Lock classes uncopyable, make ReadLock and WriteLock moveable.

ahupowerdns opened the pull request:
### Short description
Right now it is possible to copy our Lock classes, after which bad things will happen. Both destructors will attempt to unlock the mutex. 

With this PR 1) copying becomes impossible 2) the ReadLock and WriteLock classes gain move support, which enables storing them in containers with expected semantics. This is good for having "thousands of locks" that exhibit RAII behaviour.
### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled and tested this code
- [ ] included documentation (including possible behaviour changes)
- [X] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)
- [ ] <!-- remove this line if your PR is against master --> checked that this code was merged to master

rgacogne commented:
Alternatively, perhaps `ReadLock::upgrade()` should just go since it doesn't appear to be used and is quite misleading, because it does release the existing read lock before acquiring a write one.

ahupowerdns commented:
I have added some tests and removed the upgrade() feature which was stupid. gotIt() will give you a nonsense answer on moved locks, but I'm told you should not be doing that.
